include/revng/Model/Type.h

```diff
@@ -0,0 +1,699 @@
+
+public:
+  template<FlatCType T>
+  bool operator==(const T &Other) const {
+    return key() == Other.key();
+  }
+
+  template<FlatCType T>
+  bool operator<(const T &Other) const {
+    return key() < Other.key();
+  }
+
+  bool verify() const debug_function;
+};
+INTROSPECTION_NS(model, Type, Kind, ID, Name);
+
+namespace model {
+
+using UpcastableType = UpcastablePointer<model::Type>;
+
+template<size_t I = 0>
+inline model::UpcastableType
+makeTypeWithID(model::TypeKind::Values Kind, uint64_t ID);
+
+} // end namespace model
+
+template<>
+struct llvm::yaml::ScalarTraits<model::Type::Key>
+  : CompositeScalar<model::Type::Key, '-'> {};
+
+// Make UpcastablePointers to model types usable in KeyedObject containers
+template<>
+struct KeyedObjectTraits<model::UpcastableType> {
+  static model::Type::Key key(const model::UpcastableType &Val) {
+    return Val->key();
+  }
+  static model::UpcastableType fromKey(const model::Type::Key &K) {
+    return model::makeTypeWithID(K.first, K.second);
+  }
+};
+
+/// \brief Make UpcastableType yaml-serializable polymorphically
+template<>
+struct llvm::yaml::MappingTraits<model::UpcastableType>
+  : public PolymorphicMappingTraits<model::UpcastableType> {};
+
+/// \brief A qualifier for a model::Type
+struct model::Qualifier {
+  QualifierKind::Values Kind = QualifierKind::Invalid;
+  uint64_t Size = 0;
+
+  // Kind is not Invalid, Pointer and Const have no Size, Array has Size.
+  bool verify() const debug_function;
+
+  static Qualifier createConst() { return { QualifierKind::Const, 0 }; }
+
+  static Qualifier createPointer() { return { QualifierKind::Pointer, 0 }; }
+
+  static Qualifier createArray(uint64_t S) {
+    return { QualifierKind::Array, S };
+  }
+
+  bool isConstQualifier() const {
+    revng_assert(verify());
+    return Kind == QualifierKind::Const;
+  }
+
+  bool isArrayQualifier() const {
+    revng_assert(verify());
+    return Kind == QualifierKind::Const;
+  }
+
+  bool isPtrQualifier() const {
+    revng_assert(verify());
+    return Kind == QualifierKind::Const;
+  }
+
+  bool operator==(const Qualifier &) const = default;
+};
+INTROSPECTION_NS(model, Qualifier, Kind, Size);
+
+/// \brief Make Qualifier yaml-serializable
+template<>
+struct llvm::yaml::MappingTraits<model::Qualifier>
+  : public TupleLikeMappingTraits<model::Qualifier> {};
+
+// Make std::vector<model::Qualifier> yaml-serializable as a sequence
+LLVM_YAML_IS_SEQUENCE_VECTOR(model::Qualifier)
+
+namespace model {
+
+using TypePath = TupleTreeReference<model::Type, model::Binary>;
+
+inline model::TypePath makeTypePath(const model::Type *T) {
+  if (T) {
+    return TypePath::fromString("/Types/" + getNameFromYAMLScalar(T->key()));
+  } else {
+    return TypePath::fromString("/Types/");
+  }
+}
+
+} // end namespace model
+
+/// \brief A qualified version of a model::Type. Can have many nested qualifiers
+struct model::QualifiedType {
+  TypePath UnqualifiedType = model::makeTypePath(nullptr);
```

Comment:
Can we make `TypePath` default constructible and 
```suggestion
  TypePath UnqualifiedType;
```